### PR TITLE
feat: support for denormalized response from server

### PR DESCRIPTION
**Pull request by jimmycallin**
- [ ] I have added automatic tests where applicable
- [ ] The PR title is suitable as a release note
- [ ] The PR contains a description of what has been changed
- [ ] The description contains manual test instructions

## Changes

Adds support for denormalized response from server

## Test

Make sure the response is denormalized, that decoding is skipped as expected, and that there cannot be any cyclical references in the output.

### CODE REVIEW

File: test/session.test.ts [hunk 1]
```diff
@@ -732,6 +732,55 @@ describe("Encoding entities", () => {
...
+          id: 3,
+          __entity_type__: "Task",
+          status: {
+            __entity_type__: "Status",
+            id: 1,
+          },
+        },
+      ];
+      // @ts-ignore - Otherwise internal method used for testing purposes
+      const normalizedData = session.decode(
```
Comment (gismya) [inline test/session.test.ts:767]: Feels cleaner to do this test on the public facing level of the API, and don't think that would make the test any more complicated.

**Review by gismya — CHANGES_REQUESTED** [review]

File: source/session.ts [hunk 2]
```diff
@@ -156,6 +159,13 @@ export class Session {
       };
     }
 
+    if (denormalizeResponse) {
+      this.additionalHeaders = {
+        ...additionalHeaders,
+        "ftrack-api-options": "strict:1;denormalize:1",
```
[the remark this answers is not in the document]
Comment (gismya) [reply source/session.ts:165]: That goes for strict API mode as well

[the remark this answers is not in the document]
Comment (jimmycallin) [reply source/session.ts]: ftrack-strict-api header is deprecated and will be set as part of ftrack-api-options instead, to reduce number of headers we use (and have to update all systems to accept). we will keep it around until next major for backwards compatibility

File: source/session.ts [hunk 3]
```diff
@@ -403,8 +435,6 @@ export class Session {
       if (!dateValue.endsWith("Z") && !dateValue.includes("+")) {
```
[the remark this answers is not in the document]
Comment (jimmycallin) [reply source/session.ts]: i removed the nesting but not sure that's what you meant

File: source/session.ts [hunk 4]
```diff
@@ -194,6 +204,16 @@ export class Session {
 
     this.decodeDatesAsIso = decodeDatesAsIso;
 
+    /**
+     * By default the API server will return normalized responses, and we denormalize them in the client.
```
[the remark this answers is not in the document]
Comment (jimmycallin) [reply source/session.ts:201]: renamed to ensureSerializableResponse

**Review by gismya — APPROVED** [review]
Looks good.

**Comment by ffMathy** [comment]
Curious. What does this do? Seems interesting.

**Comment by jimmycallin** [comment]
> Curious. What does this do? Seems interesting.

by default our API returns results in a normalized manner, e.g. this query:

```
[
  {
    "action": "query",
    "expression": "select project.full_name from Task limit 2"
  }
]
```

would result in this:

```
[
  {
    "action": "query",
    "data": [
      {
        "__entity_type__": "Task",
        "id": "19d33e8e-8bff-11eb-b5d3-c2ffbce28b68",
        "project": {
          "__entity_type__": "Project",
          "id": "33b1e354-8b3d-11eb-8e99-c2ffbce28b68",
          "full_name": "Sync"
        }
      },
      {
        "__entity_type__": "Task",
        "id": "2b5dda84-8b4f-11eb-b695-c2ffbce28b68",
        "project": {
          "__entity_type__": "Project",
          "id": "33b1e354-8b3d-11eb-8e99-c2ffbce28b68",
          // NOTE - NO FULL_NAME HERE
        }
      }
    ],
    "metadata": {
      "next": {
        "offset": 2
      }
    }
  }
]
```

We are then in the javascript client recursively merging all entities with same ID to ensure all attributes are available where expected. This can in some cases lead to cyclical object references, meaning that we cannot JSON.stringify them, and things like hashing the result become a problem.

We have just added support for a denormalized response in the api which will return the result in a more expected manner:
```
[
  {
    "action": "query",
    "data": [
      {
        "__entity_type__": "Task",
        "id": "19d33e8e-8bff-11eb-b5d3-c2ffbce28b68",
        "project": {
          "__entity_type__": "Project",
          "id": "33b1e354-8b3d-11eb-8e99-c2ffbce28b68",
          "full_name": "Sync"
        }
      },
      {
        "__entity_type__": "Task",
        "id": "2b5dda84-8b4f-11eb-b695-c2ffbce28b68",
        "project": {
          "__entity_type__": "Project",
          "id": "33b1e354-8b3d-11eb-8e99-c2ffbce28b68",
          // NOTE - FULL NAME HERE AS EXPECTED
          "full_name": "Sync"
        }
      }
    ],
    "metadata": {
      "next": {
        "offset": 2
      }
    }
  }
]
```

this in turn means we don't need to recursively merge all entities, and the result is thus serializable.

Hope this explains it!

**Comment by gismya** [comment]
The default behavior of the server is to not return duplicates of the same object in the response, and instead let the API client handle the denormalization to make for example `Task.User.Name` be callable on all the tasks in the request. 

This has two side effects, that depending on the use case are bugs or features:
All the "User" with the same ID is the same javascript object
That javascript object might contain more properties than has been asked for in that specific query.

The main downside to this is that the result isn't serializable, since it might contain circular references to the javascript objects.

**Comment by ffMathy** [comment]
Thank you for that elaboration! We didn't know this, and we'd definitely prefer the non-circular one.

Very nice feature! Thanks for keeping the library up to date regularly! 😍

**Pull request merged**